core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,421 @@
-    batch_index: usize,
-    #[allow(dead_code)]
-    packet_index: usize,
+impl DeserializedPacket {
+    pub fn new(packet: Packet, bank: Option<&Arc<Bank>>) -> Result<Self, DeserializedPacketError> {
+        Self::new_internal(packet, bank, None)
+    }
+
+    #[cfg(test)]
+    fn new_with_fee_per_cu(
+        packet: Packet,
+        fee_per_cu: u64,
+    ) -> Result<Self, DeserializedPacketError> {
+        Self::new_internal(packet, None, Some(fee_per_cu))
+    }
+
+    pub fn new_internal(
+        packet: Packet,
+        bank: Option<&Arc<Bank>>,
+        fee_per_cu: Option<u64>,
+    ) -> Result<Self, DeserializedPacketError> {
+        let versioned_transaction: VersionedTransaction =
+            limited_deserialize(&packet.data[0..packet.meta.size])?;
+        let message_bytes = packet_message(&packet)?;
+        let message_hash = Message::hash_raw_message(message_bytes);
+        let is_simple_vote = packet.meta.is_simple_vote_tx();
+
+        let fee_per_cu = fee_per_cu.unwrap_or_else(|| {
+            bank.as_ref()
+                .map(|bank| compute_fee_per_cu(&versioned_transaction.message, bank))
+                .unwrap_or(0)
+        });
+        Ok(Self {
+            immutable_section: Rc::new(ImmutableDeserializedPacket {
+                original_packet: packet,
+                versioned_transaction,
+                message_hash,
+                is_simple_vote,
+                fee_per_cu,
+            }),
+            forwarded: false,
+        })
+    }
+
+    pub fn immutable_section(&self) -> &Rc<ImmutableDeserializedPacket> {
+        &self.immutable_section
+    }
 }
 
-/// Currently each banking_stage thread has a `UnprocessedPacketBatches` buffer to store
-/// PacketBatch's received from sigverify. Banking thread continuously scans the buffer
-/// to pick proper packets to add to the block.
-#[derive(Default)]
-pub struct UnprocessedPacketBatches(VecDeque<DeserializedPacketBatch>);
+impl PartialOrd for DeserializedPacket {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
 
-impl std::ops::Deref for UnprocessedPacketBatches {
-    type Target = VecDeque<DeserializedPacketBatch>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl Ord for DeserializedPacket {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self
+            .immutable_section()
+            .fee_per_cu()
+            .cmp(&other.immutable_section().fee_per_cu())
+        {
+            Ordering::Equal => self
+                .immutable_section()
+                .sender_stake()
+                .cmp(&other.immutable_section().sender_stake()),
+            ordering => ordering,
+        }
     }
 }
 
-impl std::ops::DerefMut for UnprocessedPacketBatches {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct PacketPriorityQueueEntry {
+    fee_per_cu: u64,
+    sender_stake: u64,
```

Comment:
Is the reason for copying `fee_per_cu` and `sender_stake` out from `ImmutableDeserializedPacket` to avoid deref during `cmp`op?

Reply:
ah, this was before i moved the `packet_ref: Rc<ImmutableDeserializedPacket>,` into this struct, before it was just the `Packet`. I think i can now remove the `PacketPriorityQueueEntry` entirely and replace it with `ImmutableDeserializedPacket`